internal/pkg/scorecard/scorecard.go

```diff
@@ -274,38 +303,71 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
 		suites = append(suites, olmTests)
 	}
 	totalScore := 0.0
+	// Update the state for the tests
 	for _, suite := range suites {
-		fmt.Printf("%s:\n", suite.GetName())
-		for _, result := range suite.TestResults {
-			fmt.Printf("\t%s: %d/%d\n", result.Test.GetName(), result.EarnedPoints, result.MaximumPoints)
+		for _, res := range suite.TestResults {
+			res.UpdateState()
 		}
-		totalScore += float64(suite.TotalScore())
 	}
-	totalScore = totalScore / float64(len(suites))
-	fmt.Printf("\nTotal Score: %.0f%%\n", totalScore)
-	// Print suggestions
-	for _, suite := range suites {
-		for _, result := range suite.TestResults {
-			for _, suggestion := range result.Suggestions {
-				// 33 is yellow (specifically, the same shade of yellow that logrus uses for warnings)
-				fmt.Printf("\x1b[%dmSUGGESTION:\x1b[0m %s\n", 33, suggestion)
+	if viper.GetString(OutputFormatOpt) == "human-readable" {
+		for _, suite := range suites {
+			fmt.Printf("%s:\n", suite.GetName())
+			for _, result := range suite.TestResults {
+				fmt.Printf("\t%s: %d/%d\n", result.Test.GetName(), result.EarnedPoints, result.MaximumPoints)
 			}
+			totalScore += float64(suite.TotalScore())
 		}
-	}
-	// Print errors
-	for _, suite := range suites {
-		for _, result := range suite.TestResults {
-			for _, err := range result.Errors {
-				// 31 is red (specifically, the same shade of red that logrus uses for errors)
-				fmt.Printf("\x1b[%dmERROR:\x1b[0m %s\n", 31, err)
+		totalScore = totalScore / float64(len(suites))
+		fmt.Printf("\nTotal Score: %.0f%%\n", totalScore)
+		// Print suggestions
+		for _, suite := range suites {
+			for _, result := range suite.TestResults {
+				for _, suggestion := range result.Suggestions {
+					// 33 is yellow (specifically, the same shade of yellow that logrus uses for warnings)
+					fmt.Printf("\x1b[%dmSUGGESTION:\x1b[0m %s\n", 33, suggestion)
+				}
+			}
+		}
+		// Print errors
+		for _, suite := range suites {
+			for _, result := range suite.TestResults {
+				for _, err := range result.Errors {
+					// 31 is red (specifically, the same shade of red that logrus uses for errors)
+					fmt.Printf("\x1b[%dmERROR:\x1b[0m %s\n", 31, err)
+				}
 			}
 		}
 	}
+	if viper.GetString(OutputFormatOpt) == "json" {
+		capturedLog := make(chan string)
+		// copy the output in a separate goroutine so printing can't block indefinitely
+		go func() {
+			var buf bytes.Buffer
+			io.Copy(&buf, readLog)
+			capturedLog <- buf.String()
+		}()
+
+		// back to normal state
+		writeLog.Close()
+		os.Stdout = origStdout // restoring the real stdout
+		log := <-capturedLog
```

Comment:
If we need to keep the stdout/stderr/log capture code, can we extract it into a function? It looks like lines 342-353 are the same as 105-116.

Reply:
I'll use your suggestion for using `log.SetOutput` and `fmt.Fprintf(buf, ...)`, so this should be much simplified then.